system-helper/flatpak-system-helper.c

```diff
@@ -51,6 +59,254 @@ G_DEFINE_AUTOPTR_CLEANUP_FUNC (AutoPolkitAuthorizationResult, g_object_unref)
+  gchar *commit;
+
+  /* root_dir_name points to root owned cache directory for the child repo.
+   * Actual file creation of the ongoing pull happens in this directory.
+   * Syntax: $repo/tmp/flatpak-cache-$ref-XXXXXX
+   *
+   * user_dir_name points to cache directory which is bind mount to root_dir_name
+   * using bindfs. This is given back to the client to initiate the pull.
+   * Syntax: $root_dir_name-bindfs
+   */
+  gchar *root_dir_name;
+  gchar *user_dir_name;
+  gchar *child_repo_path;
+  gboolean preserve; /* Don't rm -rf the child pull. It can be used to resume a partial pull */
+
+  GUnixMountMonitor *mount_monitor;
+  gulong mount_monitor_id;
+  guint unsuccessful_mount_timeout_id;
+  GSubprocess *bindfs;
+} OngoingPull;
+
+static void
+ongoing_pull_free (OngoingPull *pull)
+{
+  g_autoptr(GFile) root_dir = NULL;
+  g_autoptr(GFile) user_dir = NULL;
+  g_autoptr(GError) local_error = NULL;
+
+  root_dir = g_file_new_for_path (pull->root_dir_name);
+  user_dir = g_file_new_for_path (pull->user_dir_name);
+
+  g_cancellable_cancel (pull->cancellable);
+
+  g_clear_object (&pull->bindfs);
+
+  if (!pull->preserve)
+    {
+      if (!flatpak_rm_rf (root_dir, NULL, &local_error))
+        {
+          g_warning ("Unable to remove ongoing pull's root dir at: %s", local_error->message);
+          g_clear_error (&local_error);
+        }
+
+      if (!flatpak_rm_rf (user_dir, NULL, &local_error))
+        {
+          g_warning ("Unable to remove ongoing pull's user dir at: %s", local_error->message);
+          g_clear_error (&local_error);
+        }
+    }
+
+  g_clear_object (&pull->mount_monitor);
+
+  g_clear_pointer (&pull->root_dir_name, g_free);
+  g_clear_pointer (&pull->user_dir_name, g_free);
+  g_clear_pointer (&pull->child_repo_path, g_free);
+  g_clear_pointer (&pull->ref, g_free);
+  g_clear_pointer (&pull->commit, g_free);
+
+  g_slice_free (OngoingPull, pull);
+}
+
+/* This makes sure that we terminate the bindfs mount which in turn
+   releases the ref taken by the call to g_subprocess_wait_check_async. */
+static void
+ongoing_pull_teardown (OngoingPull *pull)
+{
+  if (pull->watch_id)
+    g_clear_handle_id (&pull->watch_id, g_bus_unwatch_name);
+
+  if (pull->bindfs != NULL)
+    {
+      /* We are probably here because bindfs was not terminated during deploy or we are on error
+         code path and trying to cleanup the mount. Hence, we try force unmount our bindfs directory,
+         to make sure no open FDs can write to it anymore. If this fails, we kill the subprocess and
+         try to cleanup /etc/mtab mount entry. */
+      if (umount2 (pull->user_dir_name, MNT_FORCE) == -1)
+        {
+          g_autoptr(GError) local_error = NULL;
+          glnx_throw_errno_prefix (&local_error, "Failed to unmount bindfs dir %s", pull->user_dir_name);
+          g_warning ("%s; killing subprocess", local_error->message);
+          g_clear_error (&local_error);
+
+          g_subprocess_force_exit (pull->bindfs);
```

Comment:
I think we should force-exit the fuse daemon even in a successful umount case, because the mount may live on in another namespace.